### ISSUE TITLE
decodeUtf8 appears to have a memory leak. 

### ISSUE THREAD

**Issue by michaelochurch** [bos/text#123]
This piece of code has a memory leak, and as far as I can tell, it shouldn't. 

I had suspected that it was a failure of laziness in decodeUtf8, but Michael Snoyman disproved this with a counterexample: https://gist.github.com/snoyberg/bfda1d0aac3f96c9d455 . The semantics are lazy. The performance problem exists, however. 

```haskell
import qualified Data.Text.Lazy as LazyText
import qualified Data.ByteString.Lazy.Char8 as LazyBytes
import Data.Text.Lazy.Encoding (decodeUtf8)

main :: IO ()
main = do
  let filename = "resources/bigfiles/large.txt" -- 5 GB text file.
  bytes <- LazyBytes.readFile filename -- verified OK (doesn't force a full read)
  let text = decodeUtf8 bytes  -- is a Data.Text.Lazy.Text
  print $ LazyText.last text   -- shouldn't blow heap (doesn't "hold its head") but does.
```

When run like so, it blows the heap. This indicates that the already-processed parts of the Lazy.Text (that can be thrown away if we're calling Lazy.Text.last) are being held in memory. 

```
$ dist/build/go/go +RTS -M120m -hc -prof -RTS resources/bigfiles/large.txt
go: Heap exhausted;
Current maximum heap size is 125829120 bytes (120 MB);
use `+RTS -M<size>' to increase it.
```

Versions are:
* base: 4.7.0.1
* bytestring: 0.10.4.0
* text: 1.2.0.4

**Comment by michaelochurch** [comment]
This repository contains a self-contained reproduction of the bug: https://github.com/michaelochurch/encodeutf8-memory-leak

**Comment by tomjaguarpaw** [comment]
Here's the culprit: https://github.com/tomjaguarpaw/text/commit/184241641cf9e008a5d54b87d0136b85e298bfe4

**Comment by tomjaguarpaw** [comment]
It's *very* surprising this hasn't been noticed before.

**Comment by michaelochurch** [comment]
Thanks!

**Comment by bos** [comment]
Nice work, @michaelochurch and @tomjaguarpaw!

**Comment by bos** [comment]
Released as 1.2.0.6
